x-pack/plugins/security_solution/server/endpoint/services/actions.ts

```diff
@@ -211,37 +256,79 @@ export const getPendingActionCounts = async (
           query: {
             bool: {
               filter: [
-                { terms: { action_id: actionIDs } }, // get results for these actions
-                { terms: { agent_id: agentIDs } }, // ignoring responses from agents we're not looking for
+                { terms: { action_id: actionIds } }, // get results for these actions
+                { terms: { agent_id: agentIds } }, // ONLY responses for the agents we are interested in (ignore others)
               ],
             },
           },
         },
       },
       { ignore: [404] }
     )
-    .then((result) => result.body?.hits?.hits?.map((a) => a._source!) || []);
-
-  // respond with action-count per agent
-  const pending: EndpointPendingActions[] = agentIDs.map((aid) => {
-    const responseIDsFromAgent = responses
-      .filter((r) => r.agent_id === aid)
-      .map((r) => r.action_id);
-    return {
-      agent_id: aid,
-      pending_actions: recentActions
-        .filter((a) => a.agents.includes(aid) && !responseIDsFromAgent.includes(a.action_id))
-        .map((a) => a.data.command)
-        .reduce((acc, cur) => {
-          if (cur in acc) {
-            acc[cur] += 1;
-          } else {
-            acc[cur] = 1;
-          }
-          return acc;
-        }, {} as EndpointPendingActions['pending_actions']),
-    };
-  });
+    .then((result) => result.body?.hits?.hits?.map((a) => a._source!) || [])
+    .catch(catchAndWrapError);
 
-  return pending;
+  if (actionResponses.length === 0) {
+    return {};
+  }
+
+  // Get the latest docs from the metadata datastream for the Elastic Agent IDs in the action responses
+  // This will be used determine if we should withhold the action id from the returned list in cases where
+  // the Endpoint might not yet have sent an updated metadata document.
+  const latestEndpointMetadataDocs = await esClient
+    .search<HostMetadata>({
+      body: {
+        query: {
+          bool: {
+            filter: [{ terms: { 'elastic.agent.id': agentIds } }],
+          },
+        },
+        collapse: {
+          field: 'elastic.agent.id',
+          inner_hits: {
+            name: 'most_recent',
+            size: 1,
+            sort: [{ 'event.created': 'desc' }],
+          },
+        },
+        aggs: {
+          total: {
+            cardinality: {
+              field: 'elastic.agent.id',
```

Comment:
aggregation might be needlessly expensive to use here. Why not use the transform destination?

If this sees a new metadata doc in the data stream, and decides the action is no longer pending, but the transform didn't run and pick it up yet, we'll get the flash, right? Because the "pending" state is shown at the output of the transform